### PR TITLE
Restrict files written by tests

### LINKED ISSUE #1060: Stop tests from writing files everywhere
### Please describe your wishes and possible alternatives to achieve the desired result.

AnnData’s tests are not well behaved and write `test.h5ad` and other things into the work directory.

We should set the working directory to read only in CI during test runs to prevent this.

### PR DESCRIPTION
**Pull request by flying-sheep**
- [x] Closes #1060
- [x] Tests added
- [x] Release note unnecessary

### CODE REVIEW

**Comment by flying-sheep** [comment]
@ivirshup implemented your suggestions from the call!

File: pyproject.toml [hunk 1]
```diff
@@ -81,6 +81,7 @@ test = [
     "loompy>=3.0.5",
     "pytest>=6.0",
     "pytest-cov>=2.10",
+    "pytest-nunit",
```
Comment (ivirshup) [inline pyproject.toml:84]: Why are we adding a dependency? Is this doing anything, and if not, can we remove this?

File: conftest.py [hunk 2]
```diff
@@ -1,15 +1,37 @@
...
+
+
+doctest_marker = pytest.mark.usefixtures("doctest_env")
+
+
+@pytest.fixture
+def doctest_env(cache: pytest.Cache, tmp_path: Path) -> None:
+    from scanpy import settings
+
+    old_dd, settings.datasetdir = settings.datasetdir, cache.mkdir("scanpy-data")
```
Comment (ivirshup) [inline conftest.py:20]: Where does the cached `scanpy-data` get made?

File: .azure-pipelines.yml [hunk 3]
```diff
@@ -37,15 +37,13 @@ jobs:
         displayName: Cache pip packages
 
       - script: |
-          python -m pip install --upgrade pip
-          pip install pytest-cov wheel
+          python -m pip install --upgrade pip wheel
```
Comment (ivirshup) [inline .azure-pipelines.yml:40]: Why is this changing? I think we intentionally did this to make sure we used an upgraded pip to install everything.

File: anndata/compat/__init__.py [hunk 4]
```diff
@@ -24,7 +29,31 @@ class Empty:
...
+        _old_cwd: list[Path] = field(default_factory=list)
+
+        def __enter__(self) -> None:
+            self._old_cwd.append(Path())
+            os.chdir(self.path)
+
+        def __exit__(self, *_exc_info) -> None:
+            os.chdir(self._old_cwd.pop())
+
+
```
Comment (ivirshup) [inline anndata/compat/__init__.py:53]: Wouldn't this make more sense in `utils` than `compat`?

**Review by ivirshup — CHANGES_REQUESTED** [review]
Can you give a summary of what goes where? Is it only coverage that's going into `test-data`?

**Comment by ivirshup** [comment]
Also, I get why you want to make the data files not be written to the root. Why move test outputs like reports and coverage from their default locations?

File: conftest.py (hunk 2 above)
Comment (flying-sheep) [reply conftest.py:20]: in .pytest-cache

File: .azure-pipelines.yml (hunk 3 above)
Comment (flying-sheep) [reply .azure-pipelines.yml:40]: still the case, this only removes `pytest-cov`

File: anndata/compat/__init__.py (hunk 4 above)
Comment (flying-sheep) [reply anndata/compat/__init__.py:53]: it’s for backwards compat with Python versions that don’t have `contextlib.chdir`.

**Comment by flying-sheep** [comment]
> Why move test outputs like reports and coverage from their default locations?

so test outputs end up in one location that can be .gitignored instead of polluting the project directory.

File: pyproject.toml [hunk 5]
```diff
@@ -107,13 +107,16 @@ source = "vcs"
 version-file = "anndata/_version.py"
 
 [tool.coverage.run]
+data_file = "test-data/coverage"
```
Comment (ivirshup) [inline pyproject.toml:110]:
```suggestion
```

File: pyproject.toml [hunk 6]
```diff
@@ -107,13 +107,16 @@ source = "vcs"
...
 source = ["anndata"]
 omit = [
 	"setup.py",
 	"versioneer.py",
 	"anndata/_version.py",
 	"**/test_*.py",
 ]
+[tool.coverage.xml]
+output = "test-data/coverage.xml"
```
Comment (ivirshup) [inline pyproject.toml:119]:
```suggestion
```

File: .gitignore [hunk 7]
```diff
@@ -2,26 +2,20 @@
...
-# Tests and coverage
-/*cache/
-/data/
-/tmp.zarr/
-test.h5ad
-test.loom
-test.zarr
-.coverage
+# Test results and coverage
+/test-data/
```
Comment (ivirshup) [inline .gitignore:18]:
```suggestion
# Test results and coverage
/test-data/
.coverage
```

**Review by ivirshup — CHANGES_REQUESTED** [review]
I don't want to change local behavior, so have made a set of suggestions reverting that. Otherwise looks good.

> documented in .gitignore:

I'm not completley clear on what counts as a "test-result".

**Review by flying-sheep — COMMENTED** [review]
&nbsp;

**Review by ivirshup — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 8]
```diff
@@ -2,26 +2,21 @@
...
-/*cache/
-/data/
-/tmp.zarr/
-test.h5ad
-test.loom
-test.zarr
-.coverage
+# Test results (nunit/junit) and coverage
+/test-data/
+*coverage*
```
[the remark this answers is not in the document]
Comment (flying-sheep) [reply .gitignore]: somewhere, yes, in the project dir, no.

**Review by ivirshup — APPROVED** [review]

**Pull request merged**